baseplate/clients/kafka.py

```diff
@@ -0,0 +1,208 @@
+from baseplate import Span
+from baseplate.clients import ContextFactory
+from baseplate.lib import config
+
+logger = logging.getLogger(__name__)
+
+TOPIC_SUFFIX = "_avro"
+CHECK_INTERVAL_SECONDS = 0.001
+
+
+class AvroProducerContextFactory(ContextFactory):
+    def __init__(
+        self,
+        bootstrap_servers: str,
+        schema_registry: str,
+        acks: int,
+        thread_pool_max_size: int = 10,
+        prefetch_schema_ids: Optional[List[int]] = None,
+    ):
+        self.avro_producer = AvroProducer(
+            {
+                "bootstrap.servers": bootstrap_servers,
+                "schema.registry.url": schema_registry,
+                "acks": acks,
+                # force sending 1 message at a time--don't wait for larger batches
+                "queue.buffering.max.ms": 0,
+                "max.in.flight": 1,
+                # hardcode the defaults here because we reference them with our manual refreshing
+                "topic.metadata.refresh.interval.ms": 300000,
+                "metadata.max.age.ms": 900000,
+            }
+        )
+        self.threadpool = ThreadPoolExecutor(thread_pool_max_size)
+        self.initialize(self.threadpool, prefetch_schema_ids or [])
+        self.last_initialized_at = time.time()
+
+    def initialize(
+        self, threadpool: ThreadPoolExecutor, prefetch_schema_ids: List[int], block: bool = True
+    ) -> None:
+        """Do some setup tasks to make sure the producer is ready to use."""
+        futures = []
+
+        # make a metadata request, otherwise we'll have to do this on our
+        # first produce which can add up to 1s of extra latency.
+        # list_topics() is a blocking call, so we have to use a separate
+        # thread to avoid blocking the gevent event loop. We can't easily
+        # make confluent_kafka gevent compatible here because it's wrapping
+        # the C library librdkafka.
+        futures.append(threadpool.submit(self.avro_producer.list_topics))
+
+        # fetch and cache all the schemas we plan to use
+        if prefetch_schema_ids:
+            for schema_id in prefetch_schema_ids:
+                futures.append(
+                    threadpool.submit(
+                        self.avro_producer._serializer.registry_client.get_by_id, schema_id
+                    )
+                )
+
+        if block:
+            _, not_done = wait(futures, timeout=1)
+            if not_done:
+                logger.info("AvroProducer timeout while initializing")
+
+    def make_object_for_context(self, name: str, span: Span) -> "KafkaAvroProducer":
+        # When initializing the producer we request the metadata by calling list_topics()
+        # Metadata is automatically refreshed every 5 minutes (topic.metadata.refresh.interval.ms)
+        # and expires from cache after 15 minutes (metadata.max.age.ms). We should attempt to
+        # refresh manually before it expires (after 10 minutes).
+        METADATA_REFRESH_INTERVAL = 600
```

Comment:
Can you make this either an object or module constant and also use:

```python
timedelta(minutes=10).total_seconds()
```

to initialize it so it's clear what the interval is and what units it is in.